File: src/main/java/com/ullgren/modern/simple3d/render/texture/MetalTexture.java
package com.ullgren.modern.simple3d.render.texture;

/**
 * Procedural metallic surface texture.
 * <p>
 * Metal is characterised by sharp specular highlights tinted by the body's own
 * colour, compressed midtones, and a narrow white sheen at the brightest point.
 * A subtle single-octave noise grain breaks up the surface uniformity.
 * <p>
 * Unlike the web version (which pre-samples noise per vertex), Java computes the
 * noise per pixel — the JIT makes this cheap enough and produces finer grain detail.
 */
public final class MetalTexture implements Texture {

  /** Minimum effective brightness — metal reflects ambient light well but lower than stone for contrast. */
  private static final double METAL_AMBIENT   = 0.20;
  /** Diffuse scale — reduced to let specular highlights dominate. */
  private static final double DIFFUSE_SCALE   = 0.45;
  /** Exponent for body-color-tinted specular. */
  private static final int    SPEC_POWER      = 6;
  /** Intensity of body-color specular. */
  private static final double SPEC_STRENGTH   = 0.55;
  /** Exponent for narrow white highlight sheen. */
  private static final int    WHITE_POWER     = 14;
  /** Intensity of white highlight. */
  private static final double WHITE_STRENGTH  = 0.20;
  /** Exponent for wide soft glare bloom around the specular peak. */
  private static final int    GLARE_POWER     = 4;
  /** Intensity of soft glare bloom. */
  private static final double GLARE_STRENGTH  = 0.18;
  /** Surface grain noise range: [GRAIN_BASE, GRAIN_BASE + GRAIN_RANGE]. */
  private static final double GRAIN_BASE      = 0.94;
  private static final double GRAIN_RANGE     = 0.06;
  /** Noise frequency — single octave for subtle metallic grain. */
  private static final double NOISE_FREQ      = 0.025;

  // Pre-computed permutation table for value noise (256-entry, doubled to avoid wrapping).
  private static final int[] PERM = buildPerm();

  @Override
  public int applyPacked(double wx, double wy, double wz,
                          int baseR, int baseG, int baseB,
                          float shade) {
    // Subtle surface grain from single-octave noise.
    double noise = valueNoise3D(wx * NOISE_FREQ, wy * NOISE_FREQ, wz * NOISE_FREQ);

    // Compressed diffuse with ambient floor.
    double metalDiffuse = Math.max(METAL_AMBIENT, shade * DIFFUSE_SCALE);

    // Body-color-tinted specular (characteristic of metallic reflection).
    double metalSpec = Math.pow(shade, SPEC_POWER) * SPEC_STRENGTH;

    // Narrow white sheen at the brightest highlights for "shine".
    double whiteSpec = Math.pow(shade, WHITE_POWER) * WHITE_STRENGTH;

    // Wide soft glare bloom — creates the characteristic polished-metal glow.
    double glare = Math.pow(shade, GLARE_POWER) * GLARE_STRENGTH;

    // Surface grain.
    double grain = GRAIN_BASE + noise * GRAIN_RANGE;

    double effective = (metalDiffuse + metalSpec) * grain;
    double whiteTotal = whiteSpec + glare;

    int r = Math.min(255, (int) (baseR * effective + 255 * whiteTotal));
    int g = Math.min(255, (int) (baseG * effective + 255 * whiteTotal));
    int b = Math.min(255, (int) (baseB * effective + 255 * whiteTotal));

    return 0xFF000000 | (r << 16) | (g << 8) | b;
  }

  // ------------------------------------------------------------------
  // 3D value noise (single octave)
  // ------------------------------------------------------------------

  private static double valueNoise3D(double x, double y, double z) {
    int ix = (int) Math.floor(x);
    int iy = (int) Math.floor(y);
    int iz = (int) Math.floor(z);

    double fx = x - ix;
    double fy = y - iy;
    double fz = z - iz;

    double ux = fade(fx);
    double uy = fade(fy);
    double uz = fade(fz);

    double v000 = latticeValue(ix,     iy,     iz    );
    double v100 = latticeValue(ix + 1, iy,     iz    );
    double v010 = latticeValue(ix,     iy + 1, iz    );
    double v110 = latticeValue(ix + 1, iy + 1, iz    );
    double v001 = latticeValue(ix,     iy,     iz + 1);
    double v101 = latticeValue(ix + 1, iy,     iz + 1);
    double v011 = latticeValue(ix,     iy + 1, iz + 1);
    double v111 = latticeValue(ix + 1, iy + 1, iz + 1);

    return lerp(uz,
        lerp(uy, lerp(ux, v000, v100), lerp(ux, v010, v110)),
        lerp(uy, lerp(ux, v001, v101), lerp(ux, v011, v111)));
  }

  private static double latticeValue(int ix, int iy, int iz) {
    int h = PERM[(PERM[(PERM[ix & 255] + iy) & 255] + iz) & 255];
    return h / 255.0;
  }

  private static double fade(double t) { return t * t * t * (t * (t * 6 - 15) + 10); }
  private static double lerp(double t, double a, double b) { return a + t * (b - a); }

  /** Builds a deterministic 256-entry permutation table (LCG seed = 137). */
  private static int[] buildPerm() {
    int[] base = new int[256];
    long rng = 137L;
    for (int i = 0; i < 256; i++) {
      rng = (rng * 1664525L + 1013904223L) & 0xFFFFFFFFL;
      base[i] = (int) (rng & 0xFF);
    }
    int[] perm = new int[512];
    for (int i = 0; i < 512; i++) perm[i] = base[i & 255];
    return perm;
  }
}

File: src/web/src/render/texture/MetalTexture.ts
import type { Texture } from './Texture.js';

/**
 * Procedural metallic surface texture.
 *
 * Metal is characterised by sharp specular highlights tinted by the body's own
 * colour, compressed midtones, and a narrow white sheen at the brightest point.
 * A subtle single-octave noise grain breaks up the surface uniformity.
 *
 * Performance: noise is evaluated once per vertex via {@link prepare} and the
 * result is interpolated linearly across each triangle by the Renderer.
 */
export class MetalTexture implements Texture {

  /** Minimum effective brightness — metal reflects ambient light well but lower than stone for contrast. */
  private static readonly METAL_AMBIENT   = 0.20;
  /** Diffuse scale — reduced to let specular highlights dominate. */
  private static readonly DIFFUSE_SCALE   = 0.45;
  /** Exponent for body-color-tinted specular. */
  private static readonly SPEC_POWER      = 6;
  /** Intensity of body-color specular. */
  private static readonly SPEC_STRENGTH   = 0.55;
  /** Exponent for narrow white highlight sheen. */
  private static readonly WHITE_POWER     = 14;
  /** Intensity of white highlight. */
  private static readonly WHITE_STRENGTH  = 0.15;
  /** Surface grain noise range: [GRAIN_BASE, GRAIN_BASE + GRAIN_RANGE]. */
  private static readonly GRAIN_BASE      = 0.94;
  private static readonly GRAIN_RANGE     = 0.06;
  /** Noise frequency — lower than stone so grain is visible after per-vertex interpolation. */
  private static readonly NOISE_FREQ      = 0.025;

  /** Deterministic 256-entry permutation table, doubled to 512 (LCG seed = 137). */
  private static readonly PERM: Uint8Array = (() => {
    const base = new Uint8Array(256);
    let rng = 137;
    for (let i = 0; i < 256; i++) {
      rng = ((rng * 1664525 + 1013904223) >>> 0);
      base[i] = rng & 0xFF;
    }
    const perm = new Uint8Array(512);
    for (let i = 0; i < 512; i++) perm[i] = base[i & 255];
    return perm;
  })();

  /** Pre-samples surface grain noise at an object-space vertex position. */
  prepare(wx: number, wy: number, wz: number): number {
    const f = MetalTexture.NOISE_FREQ;
    return MetalTexture.valueNoise3D(wx * f, wy * f, wz * f);
  }

  applyPacked(
    texValue: number,
    baseR: number, baseG: number, baseB: number,
    shade: number,
  ): number {
    // Compressed diffuse with ambient floor.
    const metalDiffuse = Math.max(MetalTexture.METAL_AMBIENT, shade * MetalTexture.DIFFUSE_SCALE);

    // Body-color-tinted specular (characteristic of metallic reflection).
    const metalSpec = Math.pow(shade, MetalTexture.SPEC_POWER) * MetalTexture.SPEC_STRENGTH;

    // Narrow white sheen at the brightest highlights for "shine".
    const whiteSpec = Math.pow(shade, MetalTexture.WHITE_POWER) * MetalTexture.WHITE_STRENGTH;

    // Subtle surface grain from interpolated noise.
    const grain = MetalTexture.GRAIN_BASE + texValue * MetalTexture.GRAIN_RANGE;

    const effective = (metalDiffuse + metalSpec) * grain;

    const r = Math.min(255, (baseR * effective + 255 * whiteSpec) | 0);
    const g = Math.min(255, (baseG * effective + 255 * whiteSpec) | 0);
    const b = Math.min(255, (baseB * effective + 255 * whiteSpec) | 0);

    return (r << 16) | (g << 8) | b;
  }

  private static valueNoise3D(x: number, y: number, z: number): number {
    const ix = Math.floor(x);
    const iy = Math.floor(y);
    const iz = Math.floor(z);
    const fx = x - ix, fy = y - iy, fz = z - iz;
    const ux = MetalTexture.fade(fx);
    const uy = MetalTexture.fade(fy);
    const uz = MetalTexture.fade(fz);

    const v000 = MetalTexture.lat(ix,     iy,     iz    );
    const v100 = MetalTexture.lat(ix + 1, iy,     iz    );
    const v010 = MetalTexture.lat(ix,     iy + 1, iz    );
    const v110 = MetalTexture.lat(ix + 1, iy + 1, iz    );
    const v001 = MetalTexture.lat(ix,     iy,     iz + 1);
    const v101 = MetalTexture.lat(ix + 1, iy,     iz + 1);
    const v011 = MetalTexture.lat(ix,     iy + 1, iz + 1);
    const v111 = MetalTexture.lat(ix + 1, iy + 1, iz + 1);

    return MetalTexture.lerp(uz,
      MetalTexture.lerp(uy, MetalTexture.lerp(ux, v000, v100), MetalTexture.lerp(ux, v010, v110)),
      MetalTexture.lerp(uy, MetalTexture.lerp(ux, v001, v101), MetalTexture.lerp(ux, v011, v111)),
    );
  }

  private static lat(ix: number, iy: number, iz: number): number {
    const perm = MetalTexture.PERM;
    return perm[(perm[(perm[ix & 255] + iy) & 255] + iz) & 255] / 255;
  }

  private static fade(t: number): number { return t * t * t * (t * (t * 6 - 15) + 10); }
  private static lerp(t: number, a: number, b: number): number { return a + t * (b - a); }
}
